4.analyze-components/scripts/nbconverted/3.within-z-svcca-analysis.py

```diff
@@ -0,0 +1,93 @@
+
+# coding: utf-8
+
+# # Determining Within Z Dimension Stability with SVCCA
+# 
+# **Gregory Way 2018**
+# 
+# Here, we apply Singular Vector Canonical Correlation Analysis ([Raghu et al. 2017](https://arxiv.org/abs/1706.05806 "SVCCA: Singular Vector Canonical Correlation Analysis for Deep Learning Dynamics and Interpretability")) ([github](https://github.com/google/svcca)) to the neuron matrices Z to quantify model stability both within and across algorithms over each bottleneck dimensionality.
+# 
+# Briefly, SVCCA uses Singular Value Decomposition (SVC) to extract the components explaining 99% of the variation.
+# This is done to remove potential dimensions described by noise.
+# Next, SVCCA performs a Canonical Correlation Analysis (CCA) on the SVD matrices to identify maximum correlations of linear combinations of both input matrices. The algorithm will identify the canonical correlations of highest magnitude across and within algorithms of the same dimensionality. 
+# 
+# The output of the SVCCA analysis is the SVCCA mean similarity score. This single number can be interpreted as a measure of similarity, or stability, in the solutions identified from each compression algorithm.
+
+# In[1]:
+
+
+import os
+import glob
+import numpy as np
+import pandas as pd
+
+from scripts.util import read_in_z, get_svcca_across_algorithm_stability
+
+
+# In[2]:
+
+
+datasets = ['TARGET', 'TCGA', 'GTEX']
+algorithms = ['pca', 'ica', 'nmf', 'dae', 'vae']
+z_dims = [2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 16, 18, 20, 25, 30,
+          35, 40, 45, 50, 60, 70, 80, 90, 100, 125, 150, 200]
+signals = (True, False)
+
+
+# In[3]:
+
+
+large_svcca_results_list = []
+for dataset in datasets:
```

Comment:
this isn't a big deal for short lists like datasets(3) and algorithms(5) but in general 3 nested for loops (or 4 nested for loops within def get_svcca_model_stability) are not a good idea. 

consider implementing zip or equivalent next time. e.g.
for dataset, z in zip(datasets, z_dims):
  do_cool_stuff(dataset, z)

Reply:
yeah... definitely agree loops are not preferred. The function works on two matrices though, and the looping is separating them out. I suppose a list comprehension may be slightly better? The bulk of the computation is occurring at the SVCCA step so I am not sure how much. Something to think about though...